README.md
# fastly-api-java
Java API Wrapper for Fastly (https://docs.fastly.com/api)

# Add snapshot repo to your maven pom.xml

```xml
<repositories>
    <repository>
      <id>sonatype snapshots</id>
      <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
    </repository>
    .....
```

# Build

```export GPG_TTY=$(tty) && mvn clean install```

# Run Integration tests

Create a file under src/test/resources called: keys.properties containing:

```
fastly.api.key=<YOUR_FASTLY_API_KEY>
fastly.service.id=<YOUR_FASTLY_SERVICE_KEY>
```

And remove the Ignore annotation from the FastlyApiClientIntegrationTest.

# how to deploy a version of this library?

http://central.sonatype.org/pages/apache-maven.html#performing-a-snapshot-deployment

# Any Issues?

Contact me at pato@split.io


CHANGES.txt
Changelog
---------
1.2.2
 Added `fastlyUrl` as an optional parameter to all methods, allowing callers to choose specific urls to hit. All methods default to `https://api.fastly.com`

1.2.1
 Added `Fastly-Soft-Purge` header, which was missing in `softPurgeKeys` method

1.2
 Include softPurgeKeys to allow batch call to purge by multiple surrogate keys.
 Add precondition check to validate we can only pass up to 256 keys.

1.1
 Include client version on each request. Example: `"User-Agent" -> "fastly-api-java-v1.1"`

1.0
 initial version

src/test/java/io/split/BaseFastlyTest.java
package io.split;

import com.google.common.base.Preconditions;
import org.junit.Before;

import java.io.IOException;
import java.util.Properties;

/**
 * Created by patricioe on 10/22/15.
 */
public class BaseFastlyTest {

    protected String _fastly_api_key;
    protected String _fastly_service_id;

    @Before
    public void before() throws IOException {
        Properties properties = new Properties();
        properties.load(this.getClass().getClassLoader().getResourceAsStream("keys.properties"));
        _fastly_api_key = properties.getProperty("fastly.api.key");
        Preconditions.checkNotNull(_fastly_api_key,
                "Fastly Key must exists. Please check that \"keys.properties\" exists in the path " +
                        "and there is entry for 'fastly.api.key'");

        _fastly_service_id = properties.getProperty("fastly.service.id");
        Preconditions.checkNotNull(_fastly_service_id,
                "Fastly Key must exists. Please check that \"keys.properties\" exists in the path " +
                        "and there is entry for 'fastly.service.id'");
    }
}


src/test/java/io/split/FastlyApiClientTest.java
package io.split;

import com.google.common.collect.Lists;
import com.ning.http.client.Response;
import io.split.fastly.client.FastlyApiClient;
import org.junit.Test;

import java.io.IOException;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.stream.Collectors;

public class FastlyApiClientTest extends BaseFastlyTest {

    @Test
    public void testPurgeAll() throws ExecutionException, InterruptedException, IOException {
        FastlyApiClient client = new FastlyApiClient(_fastly_api_key, _fastly_service_id, null);

        purgeAll(client);

        Future<Response> future = client.softPurgeKey("1446076111485");
        Response res = future.get();

        printResult(res);
    }

    private void purgeAll(FastlyApiClient client) throws InterruptedException, ExecutionException, IOException {
        Future<Response> future;
        Response res;
        future = client.purgeAll();
        res = future.get();
        printResult(res);
    }

    private void printResult(Response res) throws IOException {
        System.out.println(res.getStatusCode());
        System.out.println(res.getStatusText());
        System.out.println(res.getResponseBody());
    }

    @Test
    public void testPurgeMultipleKeys() throws ExecutionException, InterruptedException, IOException {
        FastlyApiClient client = new FastlyApiClient(_fastly_api_key, _fastly_service_id, null);

        Future<Response> future = client.softPurgeKeys(Lists.newArrayList("a", "b", "c", "d"));
        Response res = future.get();

        printResult(res);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testPurgeMoreThan256Keys() throws ExecutionException, InterruptedException, IOException {
        List<String> keys = new Random().ints(257,0,10000)
                .boxed()
                .map(i -> Integer.toString(i))
                .collect(Collectors.toList());

        FastlyApiClient client = new FastlyApiClient(_fastly_api_key, _fastly_service_id, null);

        Future<Response> future = client.softPurgeKeys(keys);
        Response res = future.get();

        printResult(res);
    }

    @Test
    public void testPurgeKey() throws ExecutionException, InterruptedException, IOException {
        FastlyApiClient client = new FastlyApiClient(_fastly_api_key, _fastly_service_id, null);

        Future<Response> future = client.softPurgeKey("a");
        Response res = future.get();

        printResult(res);
    }

}

src/test/java/io/split/fastly/client/IntegrationTest.java
package io.split.fastly.client;

import io.split.fastly.client.FastlyApiClient;
import io.split.fastly.client.VersionResolver;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;

import java.util.*;

import static org.mockito.Mockito.when;

public class IntegrationTest {

    @Test
    public void testPurgeKeys() {
        final String apikey = "someApikey";
        final String serviceId = "someServiceId";

        FastlyApiClient.AsyncHttpExecutor executor = Mockito.mock(FastlyApiClient.AsyncHttpExecutor.class);
        when(executor.execute(Mockito.any(), Mockito.any(), Mockito.any(), Mockito.any())).thenReturn(null);
        FastlyApiClient client = new FastlyApiClient(apikey, serviceId, null, executor);

        List<String> keys = Arrays.asList("key1", "key2");
        client.purgeKeys(keys, Collections.EMPTY_MAP);
        ArgumentCaptor<String> urlCaptor = ArgumentCaptor.forClass(String.class);
        ArgumentCaptor<FastlyApiClient.Method> methodCaptor = ArgumentCaptor.forClass(FastlyApiClient.Method.class);
        ArgumentCaptor<Map> headersCaptor = ArgumentCaptor.forClass(Map.class);
        ArgumentCaptor<Map> parametersCaptor = ArgumentCaptor.forClass(Map.class);
        Mockito.verify(executor).execute(urlCaptor.capture(), methodCaptor.capture(), headersCaptor.capture(),
                parametersCaptor.capture());

        assert (urlCaptor.getValue().equals(String.format("%s/service/%s/purge", FastlyApiClient.FASTLY_URL, serviceId)));
        assert (methodCaptor.getValue().equals(FastlyApiClient.Method.POST));

        Map<String, String> headers = headersCaptor.getValue();
        assert (headers.get("Fastly-Key").equals(apikey));
        assert (headers.get("Accept").equals("application/json"));
        assert (headers.get("User-Agent").equals("fastly-api-java-v"+ VersionResolver.instance().getVersion()));
        assert (headers.get("Surrogate-Key").equals(FastlyApiClient.SURROGATE_KEY_JOINER.join(keys)));
        assert (Objects.isNull(headers.get("Fastly-Soft-Purge")));
    }

    @Test
    public void testSoftPurgeKeys() {
        final String apikey = "someApikey";
        final String serviceId = "someServiceId";

        FastlyApiClient.AsyncHttpExecutor executor = Mockito.mock(FastlyApiClient.AsyncHttpExecutor.class);
        when(executor.execute(Mockito.any(), Mockito.any(), Mockito.any(), Mockito.any())).thenReturn(null);
        FastlyApiClient client = new FastlyApiClient(apikey, serviceId, null, executor);

        List<String> keys = Arrays.asList("key1", "key2");
        client.softPurgeKeys(keys);
        ArgumentCaptor<String> urlCaptor = ArgumentCaptor.forClass(String.class);
        ArgumentCaptor<FastlyApiClient.Method> methodCaptor = ArgumentCaptor.forClass(FastlyApiClient.Method.class);
        ArgumentCaptor<Map> headersCaptor = ArgumentCaptor.forClass(Map.class);
        ArgumentCaptor<Map> parametersCaptor = ArgumentCaptor.forClass(Map.class);
        Mockito.verify(executor).execute(urlCaptor.capture(), methodCaptor.capture(), headersCaptor.capture(),
                parametersCaptor.capture());

        assert (urlCaptor.getValue().equals(String.format("%s/service/%s/purge", FastlyApiClient.FASTLY_URL, serviceId)));
        assert (methodCaptor.getValue().equals(FastlyApiClient.Method.POST));

        Map<String, String> headers = headersCaptor.getValue();
        assert (headers.get("Fastly-Key").equals(apikey));
        assert (headers.get("Accept").equals("application/json"));
        assert (headers.get("User-Agent").equals("fastly-api-java-v"+ VersionResolver.instance().getVersion()));
        assert (headers.get("Surrogate-Key").equals(FastlyApiClient.SURROGATE_KEY_JOINER.join(keys)));
        assert (headers.get("Fastly-Soft-Purge").equals("1"));
    }
}


src/main/java/io/split/fastly/client/VersionResolver.java
package io.split.fastly.client;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Properties;

/**
 * read version from pom file to be used in subsequent header to identify this client.
 *
 * Created by patricioe on 5/25/16.
 */
public class VersionResolver {

    private static final Logger _log = LoggerFactory.getLogger(VersionResolver.class);

    private static VersionResolver _instance;
    private String _version;

    public String getVersion() {
        return _version;
    }

    private VersionResolver() {
        setVersion();
    }

    public static VersionResolver instance() {
        if (_instance == null) {
            synchronized (VersionResolver.class) {
                if (_instance == null) {
                    _instance = new VersionResolver();
                }
            }
        }

        return _instance;
    }

    private void setVersion() {
        Properties props = new Properties();
        try {
            props.load(this.getClass().getClassLoader().getResourceAsStream("version.properties"));
        } catch (IOException e) {
            _log.warn("Fastly Client was unable to read version to include in headers. Defaulting in 'undefined'", e);
        }
        _version = (String) props.getOrDefault("client.version", "undefined");
    }

}


src/main/java/io/split/fastly/client/FastlyApiClient.java
package io.split.fastly.client;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Joiner;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;
import com.ning.http.client.AsyncHttpClient;
import com.ning.http.client.AsyncHttpClientConfig;
import com.ning.http.client.FluentCaseInsensitiveStringsMap;
import com.ning.http.client.FluentStringsMap;
import com.ning.http.client.Response;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.Future;

import static io.split.fastly.client.FastlyApiClient.Method.POST;
import static io.split.fastly.client.FastlyApiClient.Method.PURGE;
import static io.split.fastly.client.FastlyApiClient.Method.PUT;
import static java.util.stream.Collectors.toList;

/**
 * Wrapper client for http://docs.fastly.com/api
 *
 * Created by patricioe on 10/12/15.
 */
public class FastlyApiClient {

    /* package private */ static final String FASTLY_URL = "https://api.fastly.com";
    /* package private */ static final Joiner SURROGATE_KEY_JOINER = Joiner.on(" ");

    private final Map<String, String> _commonHeaders;
    private final AsyncHttpClientConfig _config;
    private final AsyncHttpExecutor _asyncHttpExecutor;
    private final String _serviceId;
    private final String _apiKey;

    public FastlyApiClient(final String apiKey, final String serviceId) {
        this(apiKey, serviceId, null);
    }

    public FastlyApiClient(final String apiKey, final String serviceId, AsyncHttpClientConfig config) {
        this(apiKey, serviceId, config, null);
    }

    /* package private */
    @VisibleForTesting
    FastlyApiClient(final String apiKey, final String serviceId, AsyncHttpClientConfig config, AsyncHttpExecutor executor) {
        _commonHeaders = ImmutableMap.of(
                "Fastly-Key", apiKey,
                "Accept", "application/json",
                "User-Agent", "fastly-api-java-v"+ VersionResolver.instance().getVersion());
        _config = config;
        _apiKey = apiKey;
        _serviceId = serviceId;
        _asyncHttpExecutor = (Objects.isNull(executor)) ? new AsyncHttpExecutorImpl() : executor;
    }

    public Future<Response> vclUpload(int version, String vcl, String id, String name) {
        return vclUpload(version, vcl, id, name, FASTLY_URL);
    }

    public Future<Response> vclUpload(int version, String vcl, String id, String name, String fastlyUrl) {
        String apiUrl = String.format("%s/service/%s/version/%d/vcl", fastlyUrl, _serviceId, version);
        return _asyncHttpExecutor.execute(
                apiUrl,
                POST,
                ImmutableMap.<String, String> builder()
                        .putAll(_commonHeaders)
                        .put("Content-Type", "application/x-www-form-urlencoded")
                        .build(),
                ImmutableMap.<String, String> builder().put("content", vcl).put("name", name).put("id", id).build());
    }

    public List<Future<Response>> vclUpdate(int version, Map<String, String> vcl) {
        return vclUpdate(version, vcl, FASTLY_URL);
    }

    public List<Future<Response>> vclUpdate(int version, Map<String, String> vcl, String fastlyUrl) {
        return vcl.entrySet().stream().map( e -> {
            String apiUrl = String.format("%s/service/%s/version/$d/vcl/%s", fastlyUrl, _serviceId, version, e.getKey());
            return _asyncHttpExecutor.execute(
                    apiUrl,
                    PUT,
                    ImmutableMap.<String, String> builder()
                            .putAll(_commonHeaders)
                            .put("Content-Type", "application/x-www-form-urlencoded")
                            .build(),
                    ImmutableMap.<String, String> builder().put("content", e.getValue()).put("name", e.getKey()).build());

            }).collect(toList());
    }

    public Future<Response> purgeUrl(final String url) {
        return purgeUrl(url, Collections.emptyMap());
    }

    public Future<Response> softPurgeUrl(final String url) {
        return softPurgeUrl(url, Collections.emptyMap());
    }

    public Future<Response> softPurgeUrl(final String url, Map<String, String> extraHeaders) {
        return purgeUrl(url, buildHeaderForSoftPurge(extraHeaders));
    }

    public Future<Response> purgeUrl(final String url, Map<String, String> extraHeaders) {
        return _asyncHttpExecutor.execute(
                url,
                PURGE,
                ImmutableMap.<String, String>builder().putAll(_commonHeaders).putAll(extraHeaders).build(),
                Collections.emptyMap());
    }

    public Future<Response> purgeKey(String key) {
        return purgeKey(key, Collections.emptyMap());
    }

    public Future<Response> softPurgeKey(String key) {
        return softPurgeKey(key, Collections.emptyMap());
    }

    public Future<Response> softPurgeKey(String key, Map<String, String> extraHeaders) {
        return purgeKey(key, buildHeaderForSoftPurge(extraHeaders));
    }

    public Future<Response> purgeKey(String key, Map<String, String> extraHeaders) {
        return purgeKey(key, extraHeaders, FASTLY_URL);
    }

    public Future<Response> purgeKey(String key, Map<String, String> extraHeaders, String fastlyUrl) {
        String apiUrl = String.format("%s/service/%s/purge/%s", fastlyUrl, _serviceId, key);
        return _asyncHttpExecutor.execute(
                apiUrl,
                POST,
                ImmutableMap.<String, String> builder()
                        .putAll(_commonHeaders)
                        .putAll(extraHeaders)
                        .build(),
                Collections.emptyMap());
    }

    public Future<Response> purgeKeys(List<String> keys) {
        return purgeKeys(keys, FASTLY_URL);
    }

    public Future<Response> purgeKeys(List<String> keys, String fastlyUrl) {
        return purgeKeys(keys, Collections.emptyMap(), fastlyUrl);
    }

    public Future<Response> softPurgeKeys(List<String> keys) {
        return softPurgeKeys(keys, FASTLY_URL);
    }

    public Future<Response> softPurgeKeys(List<String> keys, String fastlyUrl) {
        return purgeKeys(keys, buildHeaderForSoftPurge(Collections.emptyMap()), fastlyUrl);
    }

    public Future<Response> purgeKeys(List<String> keys, Map<String, String> extraHeaders) {
        return purgeKeys(keys, extraHeaders, FASTLY_URL);
    }

    public Future<Response> purgeKeys(List<String> keys, Map<String, String> extraHeaders, String fastlyUrl) {
        Preconditions.checkNotNull(keys, "keys cannot be null!");
        Preconditions.checkArgument(keys.size() <= 256, "Fastly can't purge batches of more than 256 keys");

        String apiUrl = String.format("%s/service/%s/purge", fastlyUrl, _serviceId);
        return _asyncHttpExecutor.execute(
                apiUrl,
                POST,
                ImmutableMap.<String, String> builder()
                        .putAll(_commonHeaders)
                        .putAll(extraHeaders)
                        .put("Surrogate-Key", SURROGATE_KEY_JOINER.join(keys))
                        .build(),
                Collections.emptyMap());
    }

    public Future<Response>  purgeAll() {
        return purgeAll(FASTLY_URL);
    }

    public Future<Response>  purgeAll(String fastlyURL) {
        String apiUrl = String.format("%s/service/%s/purge_all", fastlyURL, _serviceId);
        return _asyncHttpExecutor.execute(
                apiUrl,
                POST,
                _commonHeaders,
                Collections.emptyMap());
    }

    private Map<String, String> buildHeaderForSoftPurge(Map<String, String> extraHeaders) {
        return ImmutableMap.<String, String> builder().put("Fastly-Soft-Purge", "1").putAll(extraHeaders).build();
    }

    public void closeConnectionPool() {
        _asyncHttpExecutor.close();
    }

    /**
     * Supports PURGE operations.
     */
    static class ExtendedAsyncHttpClient extends AsyncHttpClient {

        public ExtendedAsyncHttpClient(AsyncHttpClientConfig config) {
            super(config);
        }

        /**
         * Prepare an HTTP client PURGE request.
         *
         * @param url A well formed URL.
         * @return {@link BoundRequestBuilder}
         */
        public BoundRequestBuilder preparePurge(String url) {
            return requestBuilder("PURGE", url);
        }
    }


    /* package private */
    @VisibleForTesting
    interface AsyncHttpExecutor {
         Future<Response> execute( String apiUrl,Method method, Map<String, String> headers, Map<String, String> parameters);
        public void close();
    }

    /**
     * Entity Responsible for executing the requesting against the remote endpoint.
     */
    private class AsyncHttpExecutorImpl implements AsyncHttpExecutor {

        private ExtendedAsyncHttpClient client;

        private AsyncHttpClientConfig defaultConfig = new AsyncHttpClientConfig.Builder()
                .setAllowPoolingConnections(true)
                .setMaxConnections(50)
                .setMaxRequestRetry(3)
                .setMaxConnections(20000)
                .build();

        public AsyncHttpExecutorImpl() {
            client = _config != null ? new ExtendedAsyncHttpClient(_config) : new ExtendedAsyncHttpClient(defaultConfig);
        }

        public void close() {
            client.close();
        }

        public Future<Response> execute( String apiUrl,
                        Method method,
                        Map<String, String> headers,
                        Map<String, String> parameters) {

            AsyncHttpClient.BoundRequestBuilder request = getRequestForMethod(apiUrl, method);

            build(request, headers, parameters);

            return request.execute();
        }

        private AsyncHttpClient.BoundRequestBuilder getRequestForMethod(String apiURL, Method method) {

            if (method == Method.PURGE) {
                return client.preparePurge(apiURL);
            }

            if (method == Method.POST) {
                return client.preparePost(apiURL);
            }

            if (method == Method.PUT) {
                return client.preparePut(apiURL);
            }

            if (method == Method.DELETE) {
                return client.prepareDelete(apiURL);
            }

            if (method == Method.GET) {
                return client.prepareGet(apiURL);
            }

            return null;
        }

        private void build(AsyncHttpClient.BoundRequestBuilder request, Map<String, String> headers, Map<String, String> parameters) {

            FluentCaseInsensitiveStringsMap fluentCaseInsensitiveStringsMap = new FluentCaseInsensitiveStringsMap();
            headers.forEach(fluentCaseInsensitiveStringsMap::add);

            request.setHeaders(fluentCaseInsensitiveStringsMap);

            FluentStringsMap fluentStringsMap = new FluentStringsMap();
            parameters.forEach(fluentStringsMap::add);

            if (request.build().getMethod().equals("GET")) {
                request.setQueryParams(fluentStringsMap);
            } else {
                request.setFormParams(fluentStringsMap);
            }

            String host = headers.get("Host");
            if (host != null) {
                request.setVirtualHost(host);
            }
        }

    }

    enum Method {
        POST,
        PURGE,
        PUT,
        GET,
        DELETE;
    }

}


